### PR TITLE
Watch SLOs for services and message the slack thread when they start failing.

### PR DESCRIPTION
**Pull request by EvanKrall**

### CODE REVIEW

File: paasta_tools/automatic_rollbacks/signalfx.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    return float(ts) / 1000
+
+
+def tail_signalfx(query: str, lookback_seconds: float, callback: Callable) -> None:
+    if lookback_seconds > 0:
+        start_timestamp_milliseconds = (time.time() - lookback_seconds) * 1000
+    else:
+        start_timestamp_milliseconds = None
+
+    with open('/nail/home/krall/.signalfx_token') as f:
```
Comment (solarkennedy) [inline paasta_tools/automatic_rollbacks/signalfx.py]: er. I'm pretty sure we have this as a system pasta config

File: paasta_tools/automatic_rollbacks/signalfx.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+                # in SignalFX, query looks like ... .publish('<stream_label>')
+                stream_label: str = props.get('sf_streamLabel', tsid)
+
+                # sf_key seems to have the list of dimensions with a couple extra stuff
+                # The SignalFX query should minimally specify the 'service' dimension.
+                # .sum(by=['dim1','dim2'])
+                sf_key: List[str] = props.get('sf_key', [])
+
+                dimensions: Dict[str, str] = {dim: props[dim] for dim in sf_key}
+                tsid_metadata_map[tsid] = {'stream_label': stream_label, 'dimensions': dimensions}
```
Comment (solarkennedy) [inline paasta_tools/automatic_rollbacks/signalfx.py]: Is it guaranteed that we get the Metadata first? Because we use those variables in the elif

File: paasta_tools/automatic_rollbacks/slack.py [hunk 3]
```diff
@@ -69,6 +56,33 @@ def get_slack_blocks_for_deployment(
             },
         },
     ]
+    if slo_watchers is not None and len(slo_watchers) > 0:
+        num_failing = len([w for w in slo_watchers if w.failing])
```
Comment (solarkennedy) [inline paasta_tools/automatic_rollbacks/slack.py]: eventually I would like this to be a function where we can make more elaborate messages and stuff

File: paasta_tools/cli/cmds/mark_for_deployment.py [hunk 4]
```diff
@@ -979,6 +985,23 @@ def after_state_change(self):
         self.update_slack()
         super().after_state_change()
 
+    def individual_slo_callback(self, label, bad):
+        if bad:
+            message = f"SLO started failing: {label}"
+        else:
+            message = f"SLO stopped failing: {label}"
```
Comment (solarkennedy) [inline paasta_tools/cli/cmds/mark_for_deployment.py]: Can you rephrase this so it isn't a double-negative?
How about

    The {label} SLO is now OK.

File: yelp_package/extra_requirements_yelp.txt [hunk 5]
```diff
@@ -1,7 +1,20 @@
 --extra-index-url=https://pypi.yelpcorp.com/simple
+cached-property==1.3.1
 clusterman_metrics==2.2.0
 crypto_lib==4.0.0
+gitdb2==2.0.3
```
Comment (solarkennedy) [inline yelp_package/extra_requirements_yelp.txt]: Are all these other things intentional?

File: yelp_package/extra_requirements_yelp.txt [hunk 6]
```diff
@@ -1,7 +1,20 @@
...
+gitpython==2.1.9
+inflection==0.3.1
+markdown==2.4
+paasta-tools==0.84.3
```
Comment (solarkennedy) [inline yelp_package/extra_requirements_yelp.txt]: probably not this

**Comment by solarkennedy** [comment]
This looks great as a first pass.
I'm a tad worried about failures with sfx, exceptions, etc.
How do you feel about making those threads "bulletproof" so that some sort of exception in them cannot possibly make mfd crash?

File: paasta_tools/automatic_rollbacks/signalfx.py (hunk 1 above)
Comment (EvanKrall) [reply paasta_tools/automatic_rollbacks/signalfx.py]: oh, derp

File: paasta_tools/automatic_rollbacks/signalfx.py (hunk 2 above)
Comment (EvanKrall) [reply paasta_tools/automatic_rollbacks/signalfx.py]: I'm not sure. Seems like in practice it happens that way. signalfx-python also collects metadata in the same fashion, and doesn't do any sort of error handling for missing metadata: https://github.com/signalfx/signalfx-python/blob/master/signalfx/signalflow/computation.py#L74

If it's a problem, we could queue the data messages until the metadata shows up.

File: paasta_tools/cli/cmds/mark_for_deployment.py (hunk 4 above)
Comment (EvanKrall) [reply paasta_tools/cli/cmds/mark_for_deployment.py]: > This looks great as a first pass.
> I'm a tad worried about failures with sfx, exceptions, etc.
> How do you feel about making those threads "bulletproof" so that some sort of exception in them cannot possibly make mfd crash?

Exceptions in threads don't bubble up to the main thread, so we should be okay there.

**Comment by EvanKrall** [comment]
> This looks great as a first pass.
> I'm a tad worried about failures with sfx, exceptions, etc.
> How do you feel about making those threads "bulletproof" so that some sort of exception in them cannot possibly make mfd crash?

File: yelp_package/extra_requirements_yelp.txt (hunk 6 above)
Comment (EvanKrall) [reply yelp_package/extra_requirements_yelp.txt]: oops, probably not that.

File: yelp_package/extra_requirements_yelp.txt (hunk 5 above)
Comment (EvanKrall) [reply yelp_package/extra_requirements_yelp.txt]: slo-transcoder depends on gitpython which depends on gitdb2.

The intent here was to pin all the dependencies of slo-transcoder. Maybe we can get away without doing that?
